lib/livebook/notebook_manager.ex

```diff
@@ -0,0 +1,287 @@
+    recent_notebooks = update_notebook_names(state.recent_notebooks, file, name)
+    starred_notebooks = update_notebook_names(state.starred_notebooks, file, name)
+    state = %{state | recent_notebooks: recent_notebooks, starred_notebooks: starred_notebooks}
+    broadcast_changes(state, prev_state)
+    {:noreply, state, {:continue, :dump_state}}
+  end
+
+  defp remove_notebooks_on_file_system(notebook_infos, file_system_id) do
+    Enum.reject(notebook_infos, &(&1.file.file_system.id == file_system_id))
+  end
+
+  defp update_notebook_names(notebook_infos, file, name) do
+    Enum.map(notebook_infos, fn
+      %{file: ^file} = info -> %{info | name: name}
+      info -> info
+    end)
+  end
+
+  defp broadcast_changes(state, prev_state) do
+    if state.recent_notebooks != prev_state.recent_notebooks do
+      Phoenix.PubSub.broadcast(
+        Livebook.PubSub,
+        "notebook_manager:recent_notebooks",
+        {:recent_notebooks_updated, state.recent_notebooks}
+      )
+    end
+
+    if state.starred_notebooks != prev_state.starred_notebooks do
+      Phoenix.PubSub.broadcast(
+        Livebook.PubSub,
+        "notebook_manager:starred_notebooks",
+        {:starred_notebooks_updated, state.starred_notebooks}
+      )
+    end
+  end
+
+  defp load_state() do
+    attrs =
+      case Storage.fetch(@namespace, @entity_id) do
+        {:ok, attrs} -> attrs
+        _ -> %{}
+      end
+
+    file_system_by_id =
+      for file_system <- Livebook.Settings.file_systems(),
+          do: {file_system.id, file_system},
+          into: %{}
+
+    %{
+      recent_notebooks: load_notebook_infos(attrs[:recent_notebooks], file_system_by_id),
+      starred_notebooks: load_notebook_infos(attrs[:starred_notebooks], file_system_by_id)
+    }
+  end
+
+  defp load_notebook_infos(nil, _file_system_by_id), do: []
+
+  defp load_notebook_infos(notebook_infos, file_system_by_id) do
+    for %{file: file, name: name, added_at: added_at} <- notebook_infos,
+        file = load_file(file, file_system_by_id),
+        added_at = load_datetime(added_at) do
+      %{file: file, name: name, added_at: added_at}
+    end
+  end
+
+  defp load_file(%{file_system_id: file_system_id, path: path}, file_system_by_id) do
+    if file_system = file_system_by_id[file_system_id] do
+      %FileSystem.File{file_system: file_system, path: path}
+    end
+  end
+
+  defp load_datetime(datetime) do
+    {:ok, datetime, _} = DateTime.from_iso8601(datetime)
+    datetime
+  end
+
+  defp dump_state(state) do
+    attrs = [
+      recent_notebooks: dump_notebooks_infos(state.recent_notebooks),
+      starred_notebooks: dump_notebooks_infos(state.starred_notebooks)
+    ]
+
+    Storage.insert(@namespace, @entity_id, attrs)
+  end
+
+  defp dump_notebooks_infos(notebook_infos) do
+    for info <- notebook_infos do
+      %{
+        file: dump_file(info.file),
+        name: info.name,
+        added_at: DateTime.to_iso8601(info.added_at)
```

Comment:
What if we persist using from_unix to to_unix? More lightweight as it is only an integer.

Reply:
Sounds good!